state-chain/pallets/cf-threshold-signature/src/lib.rs

```diff
@@ -620,6 +1050,136 @@ pub mod pallet {
+			handle_key_ceremony_report!(
+				origin,
+				ceremony_id,
+				reported_outcome,
+				KeyRotationStatus::<T, I>::AwaitingKeygen,
+				Event::KeygenSuccessReported,
+				Event::KeygenFailureReported
+			);
+
+			Ok(().into())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(T::Weights::report_keygen_outcome())]
+		pub fn report_key_handover_outcome(
+			origin: OriginFor<T>,
+			ceremony_id: CeremonyId,
+			reported_outcome: KeygenOutcomeFor<T, I>,
+		) -> DispatchResultWithPostInfo {
+			handle_key_ceremony_report!(
+				origin,
+				ceremony_id,
+				reported_outcome,
+				KeyRotationStatus::<T, I>::AwaitingKeyHandover,
+				Event::KeyHandoverSuccessReported,
+				Event::KeyHandoverFailureReported
+			);
+
+			Ok(().into())
+		}
+
+		/// A callback to be used when the threshold signing ceremony used for keygen verification
+		/// completes.
+		///
+		/// ## Events
+		///
+		/// - [KeygenVerificationSuccess](Event::KeygenVerificationSuccess)
+		/// - [KeygenFailure](Event::KeygenFailure)
+		///
+		/// ## Errors
+		///
+		/// - [ThresholdSignatureUnavailable](Error::ThresholdSignatureUnavailable)
+		#[pallet::call_index(5)]
+		#[pallet::weight(T::Weights::on_keygen_verification_result())]
+		pub fn on_keygen_verification_result(
+			origin: OriginFor<T>,
+			keygen_ceremony_id: CeremonyId,
+			threshold_request_id: RequestId,
+			new_public_key: AggKeyFor<T, I>,
+		) -> DispatchResultWithPostInfo {
+			Self::on_key_verification_result(
+				origin,
+				threshold_request_id,
+				KeyRotationStatus::<T, I>::KeygenVerificationComplete { new_public_key },
+				Event::KeygenVerificationSuccess { agg_key: new_public_key },
+				Event::KeygenVerificationFailure { keygen_ceremony_id },
+			)
+		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(T::Weights::on_keygen_verification_result())]
+		pub fn on_handover_verification_result(
+			origin: OriginFor<T>,
+			handover_ceremony_id: CeremonyId,
+			threshold_request_id: RequestId,
+			new_public_key: AggKeyFor<T, I>,
+		) -> DispatchResultWithPostInfo {
+			Self::on_key_verification_result(
+				origin,
+				threshold_request_id,
+				KeyRotationStatus::<T, I>::KeyHandoverComplete { new_public_key },
+				Event::KeyHandoverVerificationSuccess { agg_key: new_public_key },
+				Event::KeyHandoverVerificationFailure { handover_ceremony_id },
+			)
+		}
+
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::Weights::set_keygen_response_timeout())]
+		pub fn set_keygen_response_timeout(
+			origin: OriginFor<T>,
+			new_timeout: BlockNumberFor<T>,
+		) -> DispatchResultWithPostInfo {
+			T::EnsureGovernance::ensure_origin(origin)?;
+
+			if new_timeout != KeygenResponseTimeout::<T, I>::get() {
+				KeygenResponseTimeout::<T, I>::put(new_timeout);
+				Pallet::<T, I>::deposit_event(Event::KeygenResponseTimeoutUpdated { new_timeout });
+			}
+
+			Ok(().into())
+		}
+
+		#[pallet::call_index(8)]
+		#[pallet::weight(T::Weights::set_keygen_response_timeout())]
+		pub fn set_keygen_slash_amount(
```

Comment:
We should combine all the settings extrinsics under a single one like in the validator pallet.